bgv/evaluator.go

```diff
@@ -473,106 +650,608 @@ func (eval *evaluator) mulRelin(ctIn *rlwe.Ciphertext, op1 rlwe.Operand, relin b
+		ringQ.MulCoeffsMontgomery(c00, ct1Q0.Value[1], ct2Q0.Value[1]) // c1 = c0[0]*c1[1] + c0[1]*c1[0]
+		ringQ.MulCoeffsMontgomeryThenAddLazy(c01, ct1Q0.Value[0], ct2Q0.Value[1])
+
+		ringQMul.MulCoeffsMontgomery(c00M, ct1Q1.Value[0], ct2Q1.Value[0])
+		ringQMul.MulCoeffsMontgomery(c01M, ct1Q1.Value[1], ct2Q1.Value[2])
+		ringQMul.MulCoeffsMontgomery(c00M, ct1Q1.Value[1], ct2Q1.Value[1])
+		ringQMul.MulCoeffsMontgomeryThenAddLazy(c01M, ct1Q1.Value[0], ct2Q1.Value[1])
+	}
 }
 
-// MulRelinThenAdd multiplies ctIn with op1 and adds, relinearize the result on ctOut.
-// The procedure will panic if either ctIn.Degree() or op1.Degree() > 1.
-// The procedure will panic if either ctIn == ctOut or op1 == ctOut.
-func (eval *evaluator) MulRelinThenAdd(ctIn *rlwe.Ciphertext, op1 rlwe.Operand, ctOut *rlwe.Ciphertext) {
-	eval.mulRelinThenAdd(ctIn, op1, true, ctOut)
+func (eval Evaluator) quantize(level, levelQMul int, c2Q1, c2Q2 ring.Poly) {
+
+	ringQ, ringQMul := eval.parameters.RingQ().AtLevel(level), eval.parameters.RingQMul().AtLevel(levelQMul)
+
+	// Applies the inverse NTT to the ciphertext, scales down the ciphertext
+	// by t/q and reduces its basis from QP to Q
+
+	ringQ.INTTLazy(c2Q1, c2Q1)
+	ringQMul.INTTLazy(c2Q2, c2Q2)
+
+	// Extends the basis Q of ct(x) to the basis P and Divides (ct(x)Q -> P) by Q
+	eval.basisExtenderQ1toQ2.ModDownQPtoP(level, levelQMul, c2Q1, c2Q2, c2Q2) // QP / Q -> P
+
+	// Centers ct(x)P by (P-1)/2 and extends ct(x)P to the basis Q
+	eval.basisExtenderQ1toQ2.ModUpPtoQ(levelQMul, level, c2Q2, c2Q1)
+
+	// (ct(x)/Q)*T, doing so only requires that Q*P > Q*Q, faster but adds error ~|T|
+	ringQ.MulScalar(c2Q1, eval.parameters.PlaintextModulus(), c2Q1)
+
+	ringQ.NTT(c2Q1, c2Q1)
 }
 
-func (eval *evaluator) mulRelinThenAdd(ctIn *rlwe.Ciphertext, op1 rlwe.Operand, relin bool, ctOut *rlwe.Ciphertext) {
+// MulThenAdd multiplies op0 with op1 using standard tensoring and without relinearization, and adds the result on opOut.
+// The procedure will return an error if either op0.Degree() or op1.Degree() > 1.
+// The procedure will return an error if either op0 == opOut or op1 == opOut.
+//
+// inputs:
+//   - op0: an *rlwe.Ciphertext
+//   - op1:
+//   - rlwe.ElementInterface[ring.Poly]
+//   - *big.Int, uint64, int64, int
+//   - []uint64 or []int64 (of size at most N where N is the smallest integer satisfying PlaintextModulus = 1 mod 2N)
+//   - opOut: an *rlwe.Ciphertext
+//
+// If op1 is an rlwe.ElementInterface[ring.Poly] and opOut.Scale != op1.Scale * op0.Scale, then a scale matching operation will
+// be automatically carried out to ensure that addition is performed between operands of the same scale.
+// This scale matching operation will increase the noise by a small factor.
+// For this reason it is preferable to ensure that opOut.Scale == op1.Scale * op0.Scale when calling this method.
+func (eval Evaluator) MulThenAdd(op0 *rlwe.Ciphertext, op1 rlwe.Operand, opOut *rlwe.Ciphertext) (err error) {
```

Comment:
explain in which cases it is preferable to call this instead of Mul + Add (unless this is just for convenience)

Reply:
This is for convenience, it manages the scales automatically and avoid having to use an extra buffer to store the intermediate results.